docs/specification/2025-03-26/basic/authorization.md

```diff
@@ -4,242 +4,391 @@ type: docs
+    MC->>IDP: Exchange auth code for tokens
+    IDP-->>MC: Return tokens
+    MC->>MC: Cache tokens
+    Note over MC, MS: Begin standard MCP message exchange
 ```
 
 ### 2.3 Server Metadata Discovery
 
+Servers **MUST** return the issuer identifier in `WWW-Authenticate` headers in a 
+*HTTP 401 Unauthorized* response.
+
+For servers that are overriding identity provider capabilities, such as those for token exchange or 
+dynamic client registration, they **MUST** return their fully-qualified domain as the issuer in 
+`WWW-Authenticate` headers in a *HTTP 401 Unauthorized* response.
+
 For server capability discovery:
 
-- MCP clients _MUST_ follow the OAuth 2.0 Authorization Server Metadata protocol defined
-  in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414).
-- MCP server _SHOULD_ follow the OAuth 2.0 Authorization Server Metadata protocol.
-- MCP servers that do not support the OAuth 2.0 Authorization Server Metadata protocol,
-  _MUST_ support fallback URLs.
+- MCP clients **MUST** parse the `WWW-Authenticate` response to discover the issuer URL, and construct 
+  the metadata endpoint from that issuer identifier.  
+- MCP servers **MUST** use an established identity provider and include its issuer in 
+  *HTTP 401 Unauthorized* responses.  
+- MCP clients **SHOULD** evaluate or enable the user to evaluate the trustworthiness of identity 
+  providers.  
+- MCP servers with partial support for other identity providers or that implement their own 
+  authentication and authorization logic **SHOULD** follow the OAuth 2.0 Authorization Server Metadata 
+  protocol.  
+- MCP servers **MUST** respond with a `WWW-Authenticate` header with `issuer`, `scope`, and `resource` 
+  identifiers.
 
-The discovery flow is illustrated below:
+The discovery flow is illustrated below:  
 
 ```mermaid
 sequenceDiagram
-    participant C as Client
-    participant S as Server
-
-    C->>S: GET /.well-known/oauth-authorization-server
-    alt Discovery Success
-        S->>C: 200 OK + Metadata Document
-        Note over C: Use endpoints from metadata
-    else Discovery Failed
-        S->>C: 404 Not Found
-        Note over C: Fall back to default endpoints
-    end
-    Note over C: Continue with authorization flow
+    participant MC as MCP Client
+    participant MS as MCP Server
+    participant IDP as Identity Provider
+
+    MC->>MS: Request data
+    MS-->>MC: 401 Unauthorized with WWW-Authenticate
+    MC->>MC: Construct metadata endpoint from issuer identifier
+
+    Note over MC: Continue with authorization flow
 ```
 
 #### 2.3.1 Server Metadata Discovery Headers
 
-MCP clients _SHOULD_ include the header `MCP-Protocol-Version: <protocol-version>` during
-Server Metadata Discovery to allow the MCP server to respond based on the MCP protocol
-version.
+MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server 
+Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.
```

Comment:
We have clients telling the server their version, but does the server also need to tell the client their version at any point?

Reply:
Moving versioning out of [initialization](https://spec.modelcontextprotocol.io/specification/2025-03-26/basic/lifecycle/#initialization) and into a header could make sense, but it feels like it should be a separate PR since this authorization proposal doesn't directly depend on it (that I can see).

Reply:
I don't think it should be moved out of initialization, but indicating the version early on in the authorization process would allow future changes to be made in a more compatible way.

Reply:
It does seem like something that is outside the scope of this specification to be decided (since implications go beyond auth). @dsp-ant - what do you think?